File: ourcircle/php/src/Sms.php
<?php
declare(strict_types=1);

/** Twilio SMS for Family Shield Pro circle invites, alerts, and inbound checks. Not InPmnt. */
final class Sms
{
    public static function configured(): bool
    {
        $sid = trim(Env::get('TWILIO_ACCOUNT_SID'));
        $token = trim(Env::get('TWILIO_AUTH_TOKEN'));
        $from = trim(Env::get('TWILIO_FROM'));
        if (str_contains($sid, '...') || str_contains($token, '...') || str_contains($from, '...')) {
            return false;
        }
        return str_starts_with($sid, 'AC') && strlen($token) >= 16 && str_starts_with($from, '+');
    }

    public static function normalizePhone(string $raw): string
    {
        $s = trim($raw);
        if ($s === '') {
            return '';
        }
        $digits = Analyze::digitsOnly($s);
        if (strlen($digits) === 10) {
            return '+1' . $digits;
        }
        $all = preg_replace('/\D+/', '', $s) ?? '';
        if (str_starts_with($s, '+') && strlen($all) >= 10 && strlen($all) <= 15) {
            return '+' . $all;
        }
        return '';
    }

    public static function inviteBody(string $join, string $inviter = ''): string
    {
        $who = trim($inviter) !== '' ? trim($inviter) : 'Your family';
        return 'Family Shield Pro: ' . $who . " invited you to their circle.\n"
            . $join . "\n"
            . 'Tap the link to join. ' . Analyze::CORE_RULE . ' Reply STOP to opt out.';
    }

    public static function alertBody(string $name, string $checkUrl): string
    {
        return 'PLEASE CALL ' . $name . " before they pay.\n"
            . $checkUrl . "\n"
            . 'Tap the link to open the check. ' . Analyze::CORE_RULE . ' Reply STOP to opt out.';
    }

    public static function checkBody(string $title, string $checkUrl): string
    {
        return 'OurCircle: ' . $title . "\n"
            . $checkUrl . "\n"
            . 'Tap the link. ' . Analyze::GUIDANCE . ' ' . Analyze::CORE_RULE;
    }

    public static function classifyInbound(string $body): string
    {
        $text = strtoupper(trim($body));
        if (in_array($text, ['STOP', 'STOPALL', 'UNSUBSCRIBE', 'CANCEL', 'END', 'QUIT'], true)) {
            return 'stop';
        }
        if (in_array($text, ['START', 'YES', 'UNSTOP'], true)) {
            return 'start';
        }
        if (in_array($text, ['HELP', 'INFO'], true)) {
            return 'help';
        }
        return 'check';
    }

    public static function inboundAutoReply(string $action): string
    {
        return match ($action) {
            'stop' => 'Family Shield Pro: you are opted out of SMS. Reply START to turn texts back on.',
            'start' => 'Family Shield Pro: SMS is on. Forward a sketchy text here to pause with your circle. Reply STOP to opt out.',
            'help' => 'Family Shield Pro (OurCircle): forward a suspicious text here to pause with your circle. ' . Analyze::CORE_RULE . ' Reply STOP to opt out. Email CustomerService@FamilyShieldPro.com',
            default => '',
        };
    }

    public static function twiml(string $message): string
    {
        $body = htmlspecialchars(trim($message) !== '' ? trim($message) : ' ', ENT_XML1 | ENT_QUOTES, 'UTF-8');
        return '<?xml version="1.0" encoding="UTF-8"?><Response><Message>' . $body . '</Message></Response>';
    }

    /** @param array<string,string> $params */
    public static function validSignature(string $url, array $params, string $header): bool
    {
        $token = trim(Env::get('TWILIO_AUTH_TOKEN'));
        if ($token === '' || str_contains($token, '...')) {
            return false;
        }
        ksort($params);
        $data = $url;
        foreach ($params as $k => $v) {
            $data .= $k . $v;
        }
        $expected = base64_encode(hash_hmac('sha1', $data, $token, true));
        return hash_equals($expected, trim($header));
    }

    public static function send(string $to, string $body): void
    {
        $dest = self::normalizePhone($to);
        if ($dest === '') {
            throw new RuntimeException('Need a mobile number');
        }
        if (!self::configured()) {
            throw new RuntimeException(
                'SMS is not configured. Set TWILIO_ACCOUNT_SID, TWILIO_AUTH_TOKEN, and TWILIO_FROM in .env'
            );
        }
        $sid = trim(Env::get('TWILIO_ACCOUNT_SID'));
        $token = trim(Env::get('TWILIO_AUTH_TOKEN'));
        $from = trim(Env::get('TWILIO_FROM'));
        $ch = curl_init('https://api.twilio.com/2010-04-01/Accounts/' . rawurlencode($sid) . '/Messages.json');
        curl_setopt_array($ch, [
            CURLOPT_POST => true,
            CURLOPT_USERPWD => $sid . ':' . $token,
            CURLOPT_POSTFIELDS => http_build_query([
                'To' => $dest,
                'From' => $from,
                'Body' => $body,
            ]),
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT => 30,
        ]);
        $raw = curl_exec($ch);
        $code = (int) curl_getinfo($ch, CURLINFO_HTTP_CODE);
        $err = curl_error($ch);
        curl_close($ch);
        if ($raw === false) {
            throw new RuntimeException('Twilio error: ' . $err);
        }
        if ($code >= 400) {
            throw new RuntimeException("Twilio error {$code}: {$raw}");
        }
    }
}

File: ourcircle/sms.py
"""Twilio SMS for Family Shield Pro circle invites, alerts, and inbound checks. Not InPmnt."""
from __future__ import annotations

import base64
import hashlib
import hmac
import os
import re
import urllib.error
import urllib.parse
import urllib.request
from typing import Any
from xml.sax.saxutils import escape as xml_escape

from analyze import CORE_RULE, GUIDANCE, digits_only


def sms_configured() -> bool:
    sid = (os.environ.get("TWILIO_ACCOUNT_SID") or "").strip()
    token = (os.environ.get("TWILIO_AUTH_TOKEN") or "").strip()
    frm = (os.environ.get("TWILIO_FROM") or "").strip()
    if "..." in sid or "..." in token or "..." in frm:
        return False
    return sid.startswith("AC") and len(token) >= 16 and frm.startswith("+")


def normalize_phone(raw: str) -> str:
    s = (raw or "").strip()
    if not s:
        return ""
    digits = re.sub(r"\D+", "", s)
    if len(digits) == 11 and digits.startswith("1"):
        digits = digits[1:]
    if len(digits) == 10:
        return "+1" + digits
    if s.startswith("+") and 10 <= len(digits) <= 15:
        return "+" + digits
    return ""


def invite_sms_body(join: str, inviter: str = "") -> str:
    who = (inviter or "Your family").strip() or "Your family"
    return (
        f"Family Shield Pro: {who} invited you to their circle.\n"
        f"{join}\n"
        f"Tap the link to join. {CORE_RULE} Reply STOP to opt out."
    )


def alert_sms_body(name: str, check_url: str) -> str:
    return (
        f"PLEASE CALL {name} before they pay. Open {check_url} "
        f"{CORE_RULE} Reply STOP to opt out."
    )


def check_sms_body(title: str, check_url: str) -> str:
    return f"OurCircle: {title} Open {check_url} {GUIDANCE} {CORE_RULE}"


def classify_inbound(body: str) -> str:
    text = (body or "").strip().upper()
    if text in {"STOP", "STOPALL", "UNSUBSCRIBE", "CANCEL", "END", "QUIT"}:
        return "stop"
    if text in {"START", "YES", "UNSTOP"}:
        return "start"
    if text in {"HELP", "INFO"}:
        return "help"
    return "check"


def inbound_auto_reply(action: str) -> str:
    if action == "stop":
        return "Family Shield Pro: you are opted out of SMS. Reply START to turn texts back on."
    if action == "start":
        return "Family Shield Pro: SMS is on. Forward a sketchy text here to pause with your circle. Reply STOP to opt out."
    if action == "help":
        return (
            "Family Shield Pro (OurCircle): forward a suspicious text here to pause with your circle. "
            f"{CORE_RULE} Reply STOP to opt out. Email CustomerService@FamilyShieldPro.com"
        )
    return ""


def twiml(message: str) -> str:
    body = xml_escape((message or "").strip() or " ")
    return (
        '<?xml version="1.0" encoding="UTF-8"?>'
        f"<Response><Message>{body}</Message></Response>"
    )


def valid_signature(url: str, params: dict[str, str], header: str) -> bool:
    token = (os.environ.get("TWILIO_AUTH_TOKEN") or "").strip()
    if not token or "..." in token:
        return False
    data = url
    for key in sorted(params):
        data += key + params[key]
    digest = hmac.new(token.encode("utf-8"), data.encode("utf-8"), hashlib.sha1).digest()
    expected = base64.b64encode(digest).decode("ascii")
    return hmac.compare_digest(expected, (header or "").strip())


def send_sms(*, to: str, body: str) -> dict[str, Any]:
    dest = normalize_phone(to)
    if not dest:
        raise RuntimeError("Need a mobile number")
    if not sms_configured():
        raise RuntimeError(
            "SMS is not configured. Set TWILIO_ACCOUNT_SID, TWILIO_AUTH_TOKEN, and TWILIO_FROM in .env"
        )
    sid = (os.environ.get("TWILIO_ACCOUNT_SID") or "").strip()
    token = (os.environ.get("TWILIO_AUTH_TOKEN") or "").strip()
    frm = (os.environ.get("TWILIO_FROM") or "").strip()
    payload = urllib.parse.urlencode({"To": dest, "From": frm, "Body": body or ""}).encode("utf-8")
    url = f"https://api.twilio.com/2010-04-01/Accounts/{sid}/Messages.json"
    auth = base64.b64encode(f"{sid}:{token}".encode("ascii")).decode("ascii")
    req = urllib.request.Request(
        url,
        data=payload,
        headers={
            "Authorization": f"Basic {auth}",
            "Content-Type": "application/x-www-form-urlencoded",
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            raw = resp.read().decode("utf-8")
            return {"provider": "twilio", "raw": raw}
    except urllib.error.HTTPError as exc:
        detail = exc.read().decode("utf-8", errors="replace")
        raise RuntimeError(f"Twilio error {exc.code}: {detail}") from exc
